optuna/integration/pytorch_lightning.py

```diff
@@ -69,60 +72,86 @@ def on_fit_start(self, trainer: Trainer, pl_module: "pl.LightningModule") -> Non
+            intermediate_values = _trial_system_attrs.get(_INTERMEDIATE_VALUE)
+            intermediate_values[epoch] = current_score.item()  # type: ignore[index]
+            self._trial.storage.set_trial_system_attr(
+                self._trial._trial_id, _INTERMEDIATE_VALUE, intermediate_values
+            )
+
+        # Terminate every process if any world process decides to stop.
         should_stop = trainer.strategy.broadcast(should_stop)
+        trainer.should_stop = trainer.should_stop or should_stop
         if not should_stop:
             return
 
-        if not self.is_ddp_backend:
-            message = "Trial was pruned at epoch {}.".format(epoch)
-            raise optuna.TrialPruned(message)
-        else:
-            # Stop every DDP process if global rank 0 process decides to stop.
-            trainer.should_stop = True
-            if trainer.is_global_zero:
-                self._trial.storage.set_trial_system_attr(self._trial._trial_id, _PRUNED_KEY, True)
-                self._trial.storage.set_trial_system_attr(self._trial._trial_id, _EPOCH_KEY, epoch)
-
-    def on_fit_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
-        if not self.is_ddp_backend:
-            return
+        if trainer.is_global_zero:
+            # Update system_attr from global zero process.
+            self._trial.storage.set_trial_system_attr(self._trial._trial_id, _PRUNED_KEY, True)
+            self._trial.storage.set_trial_system_attr(self._trial._trial_id, _EPOCH_KEY, epoch)
+
+    def check_pruned(self) -> None:
+        """Raise :class:`optuna.TrialPruned` manually if pruned.
+
+        Currently, intermediate_values are not properly propagated between processes due to
+        _CachedStorage. Therefore, necessary information is kept in trial_system_attrs when the
+        trial runs in a distributed situation.
+        Please do not call this method when you are not using DDP.
+        """
 
-        # Because on_validation_end is executed in spawned processes,
-        # _trial.report is necessary to update the memory in main process, not to update the RDB.
         _trial_id = self._trial._trial_id
         _study = self._trial.study
-        _trial = _study._storage._backend.get_trial(_trial_id)  # type: ignore
-        _trial_system_attrs = _study._storage.get_trial_system_attrs(_trial_id)
+        assert isinstance(_study._storage, _CachedStorage)
+        _trial_system_attrs = _study._storage._backend.get_trial_system_attrs(_trial_id)
         is_pruned = _trial_system_attrs.get(_PRUNED_KEY)
-        epoch = _trial_system_attrs.get(_EPOCH_KEY)
-        intermediate_values = _trial.intermediate_values
-        for step, value in intermediate_values.items():
-            self._trial.report(value, step=step)
+        intermediate_values = _trial_system_attrs.get(_INTERMEDIATE_VALUE)
 
+        for epoch, score in intermediate_values.items():  # type: ignore[union-attr]
```

Comment:
As talked offline, `intermediate_values` cannot be `None`.
So, how about adding an assertion instead of `type ignore` as follows?

```suggestion
        assert intermediate_values is not None
        for epoch, score in intermediate_values.items():
```